### PR TITLE
Tenant loop

### LINKED ISSUE #75: Adding same node twice creates perpetual loop
# Environment
 * OS / version: Fedora 29
 * Processor architecture: x86_64
 * TPM Manufacturer:  emulator
 * Keylime version: master

# Description 

Add a node twice creates a perpetual loop

## Expected behavior vs. actual behavior 

Keylime_verifier and keylime_tenant should report this only once, and then break / exit the loop.

## Steps to reproduce problem

Run the following command:

`keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -f somefile --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000`

Run the same again:

```
2019-03-08 13:34:45.752 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.761 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.771 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.780 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.789 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.799 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.808 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.818 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.827 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.837 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.846 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.856 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.865 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.875 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.884 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.894 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
```

```
2019-03-08 13:34:44.995 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.004 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.004 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
2019-03-08 13:34:45.013 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.023 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.023 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
2019-03-08 13:34:45.032 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.042 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.042 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
2019-03-08 13:34:45.052 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.061 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.061 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
2019-03-08 13:34:45.070 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.080 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.080 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
```

### PR DESCRIPTION
**Pull request by nabilschear**
change to tenant client api interface
    
    add to an existing node just prints an error
    added new update command that will delete and then re-add the node with
    new parameters
    delete will now confirm that the node was deleted before returning
    
fixes #75

### CODE REVIEW

**Comment by nabilschear** [comment]
@lukehinds i don't seem to be able to tag this PR or assign it.  Also this is my first foray into doing this the proper git/contributor way.  go easy on me =) . i already managed to reference the wrong issue and make a bunch of typos.  i really hate these new mac keyboards

**Comment by lukehinds** [comment]
No worries, I think the issue was that you commit message had four spaces in front of each sentence, which resulted in it being formatted as code (you can see where I removed the 4 spaces from 'fixes').

I just need to work out how this has side stepped travis :)

**Comment by nabilschear** [comment]
Ha! the reason for the spaces is that i've tried probably 5 different ways of getting this pushed. I kept copying my commit message from git log (which puts in the spaces).  I tried doing it directly to the keylime repo.  made my own fork.  then fought with ssh and https auth.  finally i found that my local repo was corrupt.  I had to go and re-pull it and start all over.

**Comment by lukehinds** [comment]
Seeing a few things wrong (unless I am mistaken)...

I see the node register:

    1552321304.57 - keylime.registrar_client - INFO - Registration activated for node D432FBB3-D2F1-4A97-9EF7-75BD81C00000.


I then try to add again:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -f somefile --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000
Using config file /etc/keylime.conf
1552321385.39 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
1552321385.39 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
1552321385.4 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x408000
1552321385.4 - keylime.tenant - INFO - vTPM PCR Mask from policy is 0x808000
1552321385.4 - keylime.tenant - DEBUG - b64_v:l4rztbOhPDgS6fkBwIvW9C/1RpZP4muNEw7JnlQpDcY=
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 745, in main
    mytenant.do_cv()
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 481, in do_cv
    raise Exception("Node %s already existed at CV.  Please use delete or update."%self.node_uuid)
Exception: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update.
1552321385.41 - keylime.tenant - ERROR - None
1552321385.41 - keylime.tenant - ERROR - Error: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update. 
```

```
1552321702.79 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
```

Good, that's as expected. Is it possible to only output the Exception (last line) and not the python stack trace (or maybe that's because its in `DEBUG` mode).

I then try to delete:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
Using config file /etc/keylime.conf
1552321767.32 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
1552321767.32 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
1552321772.39 - keylime.tenant - ERROR - Timed out waiting for delete of node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 to complete at CV
```

```
1552321767.34 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
```

File: keylime/tenant.py [hunk 1]
```diff
@@ -505,8 +503,21 @@ def do_cvdelete(self):
         if response.status_code != 200:
             logger.error("Delete command response: %d Unexpected response from Cloud Verifier."%response.status_code)
             common.log_http_response(logger,logging.ERROR,response.json())
+            return
+        
+        deleted = False
+        for _ in range(12):
+            response = tornado_requests.request("GET", "http://%s:%s/instances/%s"%(self.cloudverifier_ip,self.cloudverifier_port,self.node_uuid),context=self.context)
+            if response.status_code == 404:
```
Comment (lukehinds) [inline keylime/tenant.py]: I think you should be looking for a `200` here?

    DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000

Comment (nabilschear) [reply keylime/tenant.py]: actually that call is to get the status of the node, so i want i to show a 404 to confirm that the node's been deleted from the database.  do you see any logs on the verifier that show that the status request is being serviced?

Comment (nabilschear) [reply keylime/tenant.py]: actually the verifier doesn't log 404s i think.

Comment (nabilschear) [reply keylime/tenant.py]: https://github.com/keylime/python-keylime/blob/master/keylime/cloud_verifier_tornado.py#L114

Comment (lukehinds) [reply keylime/tenant.py]: It looks like the delete is actually failing somehow:

```
[root@localhost ~]# keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
Using config file /etc/keylime.conf
2019-03-11 17:09:29.664 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:09:29.664 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:09:34.739 - keylime.tenant - ERROR - Timed out waiting for delete of node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 to complete at CV
```

verifier:

```
2019-03-11 17:09:29.685 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
```

I then make up a UUID:

```
[root@localhost ~]# keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81CXXXXX -c delete
Using config file /etc/keylime.conf
2019-03-11 17:09:45.092 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:09:45.092 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:09:45.106 - keylime.tenant - ERROR - Delete command response: 404 Unexpected response from Cloud Verifier.
2019-03-11 17:09:45.106 - keylime.tenant - ERROR - Response code 404: instance id not found
```

```
2019-03-11 17:09:45.104 - keylime.cloudverifier - INFO - DELETE returning 404 response. instance id: D432FBB3-D2F1-4A97-9EF7-75BD81CXXXXX not found.
```

Comment (nabilschear) [reply keylime/tenant.py]: The message "Timed out waiting for delete of node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 to complete at CV" actually comes from the status check.  the made up uuid works as expected.  can you try doing an add after the the failed delete?

Comment (lukehinds) [reply keylime/tenant.py]: I think we might be able to ignore the above, this is from when a failed tenant quote state exists.

Comment (lukehinds) [reply keylime/tenant.py]: some more details:

get the node to fail tenant quote:

```
2019-03-11 17:34:34.458 - keylime.cloudverifier - WARNING - Instance D432FBB3-D2F1-4A97-9EF7-75BD81C00000 has failed tenant quote.  stopping polling
```

Now try to add it again:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -f somefile --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000
Using config file /etc/keylime.conf
2019-03-11 17:37:09.147 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:37:09.147 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:37:09.152 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x408000
2019-03-11 17:37:09.152 - keylime.tenant - INFO - vTPM PCR Mask from policy is 0x808000
2019-03-11 17:37:09.153 - keylime.tenant - DEBUG - b64_v:g6vtZCL20tNPCugfz4hGODWAf+r7yUgDzJpCmI8Se3g=
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 745, in main
    mytenant.do_cv()
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 481, in do_cv
    raise Exception("Node %s already existed at CV.  Please use delete or update."%self.node_uuid)
Exception: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update.
2019-03-11 17:37:09.162 - keylime.tenant - ERROR - None
2019-03-11 17:37:09.162 - keylime.tenant - ERROR - Error: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update. 
```

So the CV believes it exists (which is right).

So now we delete:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -f somefile --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000
Using config file /etc/keylime.conf
2019-03-11 17:37:09.147 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:37:09.147 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:37:09.152 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x408000
2019-03-11 17:37:09.152 - keylime.tenant - INFO - vTPM PCR Mask from policy is 0x808000
2019-03-11 17:37:09.153 - keylime.tenant - DEBUG - b64_v:g6vtZCL20tNPCugfz4hGODWAf+r7yUgDzJpCmI8Se3g=
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 745, in main
    mytenant.do_cv()
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 481, in do_cv
    raise Exception("Node %s already existed at CV.  Please use delete or update."%self.node_uuid)
Exception: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update.
2019-03-11 17:37:09.162 - keylime.tenant - ERROR - None
2019-03-11 17:37:09.162 - keylime.tenant - ERROR - Error: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update. 
```

I can however delete when the node has not failed quote:

```
2019-03-11 17:39:40.212 - keylime.registrar_client - INFO - Registration activated for node D432FBB3-D2F1-4A97-9EF7-75BD81C00000.
```

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
Using config file /etc/keylime.conf
2019-03-11 17:40:02.940 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:40:02.940 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:40:02.975 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from the CV
```

Comment (lukehinds) [reply keylime/tenant.py]: Some more details.

Get the node into a failed quote state:

```
2019-03-11 17:46:28.144 - keylime.cloudverifier - WARNING - Instance D432FBB3-D2F1-4A97-9EF7-75BD81C00000 has failed tenant quote.  stopping polling
```

Try to delete the node:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
Using config file /etc/keylime.conf
2019-03-11 17:50:00.290 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:50:00.290 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:50:05.409 - keylime.tenant - ERROR - Timed out waiting for delete of node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 to complete at CV
```

If its a registered node not in a failed quote state, it deletes ok:

```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
Using config file /etc/keylime.conf
2019-03-11 17:48:25.801 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-11 17:48:25.801 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-11 17:48:25.844 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from the CV
```

The verifier is not giving us much more than the 200 each time:

```
2019-03-11 17:51:46.945 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
```

Comment (nabilschear) [reply keylime/tenant.py]: yeah, the verifier just marks the node for deletion and returns 200.  I thought that this code would just immediately delete it if it was already failed
https://github.com/keylime/python-keylime/blob/7b0a5b4c57e0efd7add2cac88a7ad0eca95e9a90/keylime/cloud_verifier_tornado.py#L151-L154

I'll debug a little and see what's going on

Comment (nabilschear) [reply keylime/tenant.py]: hrm, i wasn't able to reproduce this one.  when it is in a failed state it does delete for me.

**Comment by nabilschear** [comment]
@lukehinds  try this version.  behaves differently depending on the state of the node on the verifier

**Comment by nabilschear** [comment]
@lukehinds i wasn't sure if you'd checked this latest branch for your problem when you posted this comment?
https://github.com/keylime/python-keylime/issues/75#issuecomment-471972522

let me know if you're still having trouble.

**Comment by lukehinds** [comment]
> @lukehinds i wasn't sure if you'd checked this latest branch for your problem when you posted this comment?
> [#75 (comment)](https://github.com/keylime/python-keylime/issues/75#issuecomment-471972522)
> 
> let me know if you're still having trouble.

Thanks buddy, should be testing it shortly. I want to get this one in for the next release too

**Comment by lukehinds** [comment]
hi @nabilschear 

this still fails for me

here is the complete session: https://gist.github.com/lukehinds/a624d1672628adbccf5a95a83252fb20

I am testing this out on a fresh install everytime (that passes `run_tests.sh` first).

Just let me know if you would like me to try anything else.

**Comment by jetwhiz** [comment]
Is this being run on a machine with an emulated TPM?  If so, it looks like EK cert checking is not turned off: "Error processing ek/ekcert".  This is throwing an exception which might be causing other issues, too.

**Comment by lukehinds** [comment]
Good point, will check first thing in the morning!

**Comment by lukehinds** [comment]
Apologies for wasting a few cycles. I thought my set up script was toggling off require ek, but it seems not.

So this works better for me now, the last point I have is why use `raise Exception` and not `logger.error` - when we raise an Error we get the following stack trace outputted, that kind of suggests something is broken, when its not - the code is just reporting the state (it already exists).

```
1552982503.81 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 744, in main
    mytenant.do_cv()
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tenant.py", line 481, in do_cv
    raise Exception("Node %s already existed at CV.  Please use delete or update."%self.node_uuid)
Exception: Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already existed at CV.  Please use delete or update.
```

I don't think you need to change it in this patch. I will create a new issue and we can discuss removing them all from `tenant.py`

**Pull request merged**